src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SoftReference`1.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+
+namespace CSharpSyntaxGenerator
+{
+    /// <summary>
+    /// This class behaves like <see cref="WeakReference{T}"/>, with the added constraint that the referenced object
+    /// will not be reclaimed prior to the next Generation 2 garbage collection operation.
+    /// </summary>
+    /// <remarks>
+    /// The Generation 2 constraint only applies to the current target of the soft reference. If the target is replaced
+    /// by calling <see cref="SetTarget(T)"/>, the previous target may be immediately eligible for garbage collection.
+    /// </remarks>
+    /// <typeparam name="T">The type of object referenced.</typeparam>
+    internal sealed class SoftReference<T>
+        where T : class?
+    {
+        private readonly WeakReference<T?> _weakReference = new(null);
+        private T? _strongReference;
+
+        public SoftReference(T? target)
+        {
+            Gen2GcCallback.Register(
+                static self => ((SoftReference<T>)self).OnGen2GC(),
+                this);
+
+            if (target is not null)
+                SetTarget(target);
+        }
+
+        public void SetTarget(T target)
+        {
+            // Set _weakReference prior to setting _strongReference. TryGetTarget reads them in the opposite order.
+            _weakReference.SetTarget(target);
+            if (target is not null && GC.GetGeneration(target) < 2)
+                Volatile.Write(ref _strongReference, target);
+            else
+                Volatile.Write(ref _strongReference, null);
+        }
```

Comment:
What happens if OnGen2GC is currently in flight as this is being called?  Seems like SetTarget could race with OnGen2GC and cause the strong reference to be immediately wiped out (OnGen2GC gets the current _weakReference target and sees that it's Gen2, but before it nulls out _strongReference, SetTarget runs, sets the _strongReference, and then it's immediately zero'd out by OnGen2GC continuing).  Is that ok?  It would violate the guarantee that the target is maintained until at least Gen2, but I don't know if that's meant to hold when SetTarget is used to overwrite a previous target.

Reply:
OnGen2GC uses `CompareExchange` to ensure it only overwrites a strong reference that is known to be in Gen 2. This could lead to an object _surviving_ one or more Gen 2 GC operations when it would otherwise be collected, but as anyone who's over tried to write a GC survivability test can attest this is always possible anyway.

Reply:
> OnGen2GC uses CompareExchange

Yup, I seethat now.  Somehow I read it as Exchange.